Userland/Utilities/tail.cpp

```diff
@@ -77,19 +53,59 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     TRY(Core::System::pledge("stdio rpath"));
 
     bool follow = false;
-    int line_count = DEFAULT_LINE_COUNT;
-    char const* file = nullptr;
+    size_t line_count = DEFAULT_LINE_COUNT;
+    StringView file;
 
     Core::ArgsParser args_parser;
-    args_parser.set_general_help("Print the end ('tail') of a file.");
+    args_parser.set_general_help("Print the end - the \"tail\" - of a file.");
     args_parser.add_option(follow, "Output data as it is written to the file", "follow", 'f');
     args_parser.add_option(line_count, "Fetch the specified number of lines", "lines", 'n', "number");
-    args_parser.add_positional_argument(file, "File path", "file");
+    args_parser.add_positional_argument(file, "File path", "file", Core::ArgsParser::Required::No);
     args_parser.parse(arguments);
 
-    auto f = TRY(Core::File::open(file, Core::OpenMode::ReadOnly));
-    TRY(Core::System::pledge("stdio"));
+    // Since we can't seek the standard input, we read
+    // the entire file and split on new lines.
+    if (file.is_empty() || file == "-") {
+        do {
+            auto buffer = TRY(TRY(Core::Stream::File::standard_input())->read_all(PAGE_SIZE));
+            auto lines = StringView(buffer).split_view("\n"sv, true);
+            int i = 0;
+            for (auto line : lines) {
+                if (i >= static_cast<int>(lines.size() - line_count))
+                    outln(line);
+                i++;
+            }
+            // Since we can't have FileWatchers on the standard input either,
+            // we just loop forever if the -f option was passed.
+        } while (follow);
+        return 0;
+    }
+
+    auto f = TRY(Core::Stream::File::open(file, Core::Stream::OpenMode::Read));
+    if (!follow)
+        TRY(Core::System::pledge("stdio"));
+    auto pos = TRY(find_seek_pos(*f, line_count));
+
+    auto seek_to_end = [](Core::Stream::File& file) {
+        MUST(file.seek(0, Core::Stream::SeekMode::FromEndPosition));
+    };
+
+    TRY(tail_from_pos(*f, pos));
 
-    auto pos = find_seek_pos(*f, line_count);
-    return tail_from_pos(*f, pos, follow);
+    if (follow) {
+        seek_to_end(*f);
+
+        Core::EventLoop event_loop;
+        auto watcher = TRY(Core::FileWatcher::create());
+        watcher->on_change = [&](Core::FileWatcherEvent const& event) {
+            if (event.type == Core::FileWatcherEvent::Type::ContentModified) {
+                out(MUST(f->read_all()));
+                seek_to_end(*f);
+            }
+        };
+        TRY(watcher->add_watch(file, Core::FileWatcherEvent::Type::ContentModified));
+        TRY(Core::System::pledge("stdio"));
+        event_loop.exec();
```

Comment:
You should return the result of this event loop execution to propagate signals properly.

Reply:
Specifically: If the user interrupts tail while the event loop is running via Ctrl+C, the loop exits but the function returns 0. If you return this value instead, the program's return value will indicate that it was ended with SIGINT.